userapi/src/main/java/fk/prof/userapi/api/cache/ZkLoadInfoStore.java

```diff
@@ -0,0 +1,312 @@
+package fk.prof.userapi.api.cache;
+
+import com.codahale.metrics.Counter;
+import com.google.common.io.BaseEncoding;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.protobuf.AbstractMessage;
+import com.google.protobuf.Parser;
+import fk.prof.aggregation.AggregatedProfileNamingStrategy;
+import fk.prof.metrics.Util;
+import fk.prof.userapi.proto.LoadInfoEntities.NodeLoadInfo;
+import fk.prof.userapi.proto.LoadInfoEntities.ProfileResidencyInfo;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.api.transaction.CuratorTransactionFinal;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+
+import java.nio.charset.Charset;
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * Shared storage based on zookeeper to store the mapping profile -> ip:port and maintain resource utilization counters
+ * for each node in cluster.
+ * It is used by {@link ClusterAwareCache}.
+ *
+ * zookeeper node structure:
+ * /fkprof-userapi                                                                    (session namespace)
+ * /fkprof-userapi/nodesInfo/{ip:port} -> {@link NodeLoadInfo}                                (ephemeral)
+ * /fkprof-userapi/profilesLoadStatus/{profileName} -> {@link ProfileResidencyInfo}           (ephemeral)
+ *
+ * For resource utilization counter, currently only cached profile count is stored
+ *
+ * @see NodeLoadInfo
+ * @see ProfileResidencyInfo
+ *
+ * Created by gaurav.ashok on 01/08/17.
+ */
+class ZkLoadInfoStore {
```

Comment:
have a well defined interface which this class should implement

Reply:
Pulled out non ZK specific methods to a `LoadInfoStore` interface